### ISSUE TITLE
Issue with serviceLens map and the services based on EKS [SCALA]

### ISSUE THREAD

**Issue by HashemTaheriSonos** [aws-observability/aws-otel-java-instrumentation#65]
**Describe the bug**
We are using EKS with several namespaces. The problem is that the services are not getting recognized correctly and I think the proper attributes are not getting add to the traces correctly.
I'm using ADOT java agent(1.1.0) + ADOT agent as DaemonSet (v0.11.0) + ADOT collector (v0.11.0)

firstly while all of the services have different names and `service.name` attribute, they are not getting represented, just the connection from those services are getting represented, for example, the connection from the service to the S3 or documentDB. like below:
client --->  S3
while the expected behaviour is like, client ---> my sevice(pod) name---> S3

Secondly, I expected to have some extra information and attributes by using ADOT java agent or ADOT collector which seems not working, for instance as I was seeing that the pods are getting considered as EC2 instance, I tried to add the `k8s.cluster.name=the-cluster-name` manually to the java agent. This leads to having some changes in the representation as follows:

without setting manually the k8s cluster attribute:
![EKS_without_tag](https://user-images.githubusercontent.com/80122300/127015921-07e7e060-0703-4ff6-980d-a90b0d281c3b.jpg)

After adding the attribute: (just seems that the representation icon changed and it becomes double as I just added this attribute to one of namespaces which this image was running on it, so other namespaces are still without attribute and represented as single EC2 instance (The same image running in different namespaces))
![EKS_with_tag](https://user-images.githubusercontent.com/80122300/127016055-9d25316f-c188-4de2-ba41-1aea0d031053.jpg)
![cluster-tag-in-dashboard](https://user-images.githubusercontent.com/80122300/127016498-e9b971be-2f77-4f19-97a6-1b1d7a52ad0f.jpg)

I thought maybe it's due to the fact that i'm running the collector as DaemonSet, not Sidecar, Hence I tested it with sidecar but didn't improve the situation. I would be glad to have your idea about these issues, while I'm trying to solve them.

I created this issue as it works as documentation as well, maybe I'm missing some parts or there are some compulsory attributes that should be set in case of using multi namespace based on EKS manually. in any case i will try to keep my opened issues as update as possible to my latest findings.

 thank you in advance for your help.

**Comment by anuraaga** [comment]
Hi @HashemTaheriSonos - I'm having some trouble interpreting the issue, based on the service maps. Could you check out individual traces to see what segments may have incorrect data? For example, we naturally don't expect the number of segments in a trace to increase just from changing attributes, while the service map does display this, which makes it hard to pinpoint what could be going on.

Though I'm wondering if https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/4384 is causing issues here and could help.

**Comment by mxiamxia** [comment]
If you can provide the trace raw data on the node for each branch and we can try to check it out. 

There are 2 options usually you can get the correct AWS Resource(Origin) in X-Ray console for your services. 
1. Resource Detection package in OTel SDK. I am not sure how it can be enabled with java-auto-agent. @anuraaga could help to provide some insights. :)
2. You can enable Resource Detection Processor in ADOT collector configuration.

**Comment by anuraaga** [comment]
@mxiamxia The AWS resource detectors are enabled automatically when using the ADOT java agent (not upstream Otel java agent yet though we probably will do that too). However the SDK detectors do very little for EKS so wasn't sure if they are enough here.

**Comment by HashemTaheriSonos** [comment]
@anuraaga @mxiamxia Thanks for your response, well one of my problem is exactly why the resource detectors are not working probably at agent level (@mxiamxia I didn't know we can configure it for the collector as well, I will try to do so and let you know about the result)

@anuraaga In order to describe the issue better, I created some figures which represent and simulate the same situation.
Let's consider the below architecture for our services:

![problem - Page 1](https://user-images.githubusercontent.com/80122300/127469851-f497ac0d-ea48-4171-b34f-25fc3b7caaf0.jpeg)

Some facts:
The services which are instrumented with javaagent are based on Scala.
The collector and java agent that i'm using are the ADOT one. javaagent v1.4.0, collector 0.11.0

Now by considering passing attributes like following to the java agent, we would have representation like next figure;
`OTEL_RESOURCE_ATTRIBUTES: service.name=service-1-production,service.namespace=production`
`OTEL_RESOURCE_ATTRIBUTES: service.name=service-1-test,service.namespace=test`
...

![Page 2](https://user-images.githubusercontent.com/80122300/127471043-81d10cd0-b659-48cc-8944-66c679511ef5.jpg)

Some facts about the above figure:
The `sqs` contains the traces from both individual sqs of different env. So there are two separated sqs, and if i check the traces of this sqs entity, it contains traces from both of the sqs.
The `service 3` contains the traces from service 2, the source of the trace is service 2.

At this point i was wondering why it's not recognising that it's on EKS, and also the representation is not right for sure, so tried to add `k8s.cluster.name` attribute to **only** the  test namespaces javaagent.
`OTEL_RESOURCE_ATTRIBUTES: service.name=service-1-test,service.namespace=test,k8s.cluster.name=my-eks-cluster`

By doing so the serviceLens map changed like below:

![Page 3](https://user-images.githubusercontent.com/80122300/127472616-c3486f6b-dd77-4cf0-b15e-8436569e946a.jpg)

So the traces from the TEST env, which contained this cluster_name attribute are getting represented differently and if I check the segment resources tab, we have:
```
Account id 12345678
Ec2
      Availability zone   us-x-1x
      Instance id         i-0333311332
      Instance size        xX.XXX
      Ami id              ami-08xxxxx5f3f
Xray
      Auto instrumentation   true
      Sdk                       version 1.4.1
      Sdk                       opentelemetry for java
Eks
     Cluster name              my-eks-cluster
     Pod                        null
     Container id              null
```
(I put x to cover info, otherwise the actual values are represented correctly.)
ADDITIONAL INFO: in the collector with debug mode, the spans are look like below:
```
Span #12
    Trace ID       : 61012x6x8xx13e91c0aexxx5e54a379c
    Parent ID      :
    ID             : 90faxxf5701bxbxe
    Name           : HTTP GET
    Kind           : SPAN_KIND_CLIENT
    Start time     : 2021-07-28 09:33:32.902000132 +0000 UTC
    End time       : 2021-07-28 09:33:32.92215231 +0000 UTC
    Status code    : STATUS_CODE_UNSET
    Status message :
Attributes:
     -> http.user_agent: STRING(AHC/2.1)
     -> http.flavor: STRING(1.1)
     -> net.transport: STRING(IP.TCP)
```

Hope this made it more clear and sorry as it becomes lengthy, 
I will try to add the `resourcedetection` to the collector and let you know about the effect.

**Comment by HashemTaheriSonos** [comment]
@mxiamxia Here is the raw data of the trace, (one of the traces from `service 3`):
<details><summary>Raw JSON data</summary>

```JSON
{
    "Id": "1-61xxx570-5af0e56bcd961b2xxxx843e8",
    "Duration": 0.008,
    "LimitExceeded": false,
    "Segments": [
        {
            "Id": "9d6fcxxxx96a4a22",
            "Document": {
                "id": "9d6fcxxxx96a4a22",
                "name": "service3",
                "start_time": 1627891056.654,
                "trace_id": "1-61xxx570-5af0e56bcd961b2xxxx843e8",
                "end_time": 1627891056.6622827,
                "fault": false,
                "error": false,
                "throttle": false,
                "http": {
                    "request": {
                        "url": "http://one-end-point-of-service-3",
                        "method": "GET",
                        "user_agent": "AHC/2.1",
                        "client_ip": "x.x.x.x"
                    },
                    "response": {
                        "status": 200,
                        "content_length": 0
                    }
                },
                "aws": {
                    "account_id": "xxxxx",
                    "ec2": {
                        "availability_zone": "us-x-xa",
                        "instance_id": "i-00f40xxxx7e23b4e4",
                        "instance_size": "mx.xxlarge",
                        "ami_id": "ami-0exxxxx399dbb5f3f"
                    },
                    "xray": {
                        "auto_instrumentation": true,
                        "sdk_version": "1.1.0",
                        "sdk": "opentelemetry for java"
                    }
                },
                "annotations": {
                    "otel_resource_service_namespace": "production",
                    "otel_resource_service_name": "service3-production"
                },
                "metadata": {
                    "default": {
                        "http.flavor": "1.1",
                        "otel.resource.host.name": "ip-x-x-x-x.ec2.internal",
                        "thread.name": "AsyncHttpClient-3-2",
                        "otel.resource.telemetry.auto.version": "1.1.0-aws",
                        "otel.resource.process.pid": 1,
                        "otel.resource.os.description": "Linux 5.x",
                        "otel.resource.cloud.platform": "aws_eks",
                        "otel.resource.os.type": "LINUX",
                        "otel.resource.cloud.region": "us-x-x",
                        "otel.resource.host.type": "mx.xxlarge",
                        "thread.id": 40,
                        "otel.resource.telemetry.sdk.name": "opentelemetry",
                        "net.transport": "IP.TCP",
                        "otel.resource.cloud.availability_zone": "us-x-x",
                        "otel.resource.host.image.id": "ami-0exxxx399dbb5f3f",
                        "otel.resource.process.runtime.description": "Oracle Corporation OpenJDK 64-Bit Server VM 11.0.11+9",
                        "otel.resource.process.runtime.version": "11.0.11+9",
                        "otel.resource.host.id": "i-00f40xxxx7e23b4e4",
                        "otel.resource.process.command_line": "/usr/local/openjdk-11:bin:java -javaagent:/opt/docker/bin/../aws-opentelemetry-agent/aws-opentelemetry-agent-1.1.0.jar -Dcom.sun.management.jmxremote",
                        "otel.resource.cloud.account.id": "28xxxx433075",
                        "otel.resource.process.executable.path": "/usr/local/openjdk-11:bin:java",
                        "otel.resource.telemetry.sdk.version": "1.1.0",
                        "otel.resource.process.runtime.name": "OpenJDK Runtime Environment",
                        "otel.resource.telemetry.sdk.language": "java",
                        "otel.resource.cloud.provider": "aws"
                    }
                }
            }
        }
    ]
}
```
</details>

Also as you suggested:
> 2\. You can enable [Resource Detection Processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/resourcedetectionprocessor) in ADOT collector configuration.  In your case, you need to bring your own customized OTel yaml config with the below processor enabled.

I enabled this processor at collector and currently, it's setting "cloud.platform" which leads to having the representation below in cloud watch and x-ray:
![After-processor](https://user-images.githubusercontent.com/80122300/127839366-d8006c52-0754-4281-ab28-48064ec38940.jpeg)

**Comment by HashemTaheriSonos** [comment]
UPDATE:

Found that the root issue of such behavior is the fact that the code is based on Scala and the java agent is not fully compatible with it indeed.
There is an issue with context propagation (based on multi-thread arch), Hence each time the agent creates a new trace id as it is not receiving any trace context and think it's the root span; this leads to having a separated representation for the traces.

btw adding the `resourcedetection` to the agent also improved the overall situation, so thanks for the suggestion. yet in the case of EKS, it's just considering adding very few attributes.

seems currently just manual instrumentation would work in the case of scala by using a solution like zio-telemtery. I saw a scala-agent ticket in the roadmap but I guess it will take time to have an officially supported agent for the scala.
